src/clj/flybot/auth/oauth2_reitit.clj

```
@@ -0,0 +1,132 @@
+     (get-in request [:query-params "state"])))
+
+(defn- coerce-to-int [n]
+  (if (string? n)
+    (Integer/parseInt n)
+    n))
+
+(defn- format-access-token
+  [{{:keys [access_token expires_in refresh_token id_token] :as body} :body}]
+  (-> {:token access_token
+       :extra-data (dissoc body :access_token :expires_in :refresh_token :id_token)}
+      (cond-> expires_in (assoc :expires (-> expires_in
+                                             coerce-to-int
+                                             time/seconds
+                                             time/from-now))
+              refresh_token (assoc :refresh-token refresh_token)
+              id_token (assoc :id-token id_token))))
+
+(defn- get-authorization-code [request]
+  (get-in request [:query-params "code"]))
+
+(defn- request-params [profile request]
+  {:grant_type    "authorization_code"
+   :code          (get-authorization-code request)
+   :redirect_uri  (:redirect-uri profile)})
+
+(defn- add-header-credentials [options client-id client-secret]
+  (assoc options :basic-auth [client-id client-secret]))
+
+(defn- add-form-credentials [options client-id client-secret]
+  (assoc options :form-params (-> (:form-params options)
+                                  (merge {:client_id     client-id
+                                          :client_secret client-secret}))))
+
+(defn- get-access-token
+  [{:keys [access-token-uri client-id client-secret basic-auth?]
+    :as profile}
+   request]
+  (format-access-token
+   (try
+     @(http/request
+       (cond->
+        {:as            :json
+         :accept        :json
+         :url           access-token-uri
+         :method        :post
+         :form-params   (request-params profile request)}
+         basic-auth?       (add-header-credentials client-id client-secret)
+         (not basic-auth?) (add-form-credentials client-id client-secret)))
+     (catch Exception e
+       (ex-data e)))))
```

Catch all exception normally is not a good practice, in this case, looks like just ignore it should ok for now.